core/primitives/src/congestion_info.rs

```diff
@@ -112,8 +113,47 @@ impl CongestionControl {
 
     /// Whether we can accept new transaction with the receiver set to this shard.
     pub fn shard_accepts_transactions(&self) -> bool {
-        self.congestion_level() < self.config.reject_tx_congestion_threshold
+        self.rejection_reason().is_none()
     }
+
+    /// If the shard doesn't accept new transaction, provide the reason for
+    /// extra debugging information.
+    pub fn rejection_reason(&self) -> Option<RejectTransactionReason> {
+        let incoming_congestion = self.incoming_congestion();
+        let outgoing_congestion = self.outgoing_congestion();
+        let memory_congestion = self.memory_congestion();
+        let missed_chunks_congestion = self.missed_chunks_congestion();
+
+        let congestion_level = incoming_congestion
+            .max(outgoing_congestion)
+            .max(memory_congestion)
+            .max(missed_chunks_congestion);
```

Comment:
This seems copy pasted from the congestion_level method though I'm guessing it's intentional because you need the intermediate values as well?